File: rewrite-javascript/rewrite/src/javascript/whitespace-reconciler.ts
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {JavaScriptVisitor} from './visitor';
import {J} from '../java';
import {Cursor, Tree} from "../tree";
import {produce} from "immer";

/**
 * A visitor that reconciles whitespace from a formatted tree into the original tree.
 * Walks both trees in parallel and copies whitespace (prefix, before, after) from
 * the formatted tree to the original.
 *
 * The result preserves the original AST's structure, types, and markers while
 * applying the formatted tree's whitespace.
 *
 * When a target subtree is specified, the reconciler only applies whitespace changes
 * to that subtree and its descendants, leaving surrounding code unchanged.
 */
export class WhitespaceReconcilerVisitor extends JavaScriptVisitor<J> {
    /**
     * Cursor tracking the current position in the formatted tree.
     */
    protected formattedCursor?: Cursor;

    /**
     * Flag indicating whether the trees have compatible structure.
     */
    protected compatible: boolean = true;

    /**
     * The ID of the subtree to reconcile. If undefined, reconcile everything.
     */
    private targetSubtreeId?: string;

    /**
     * Tracks the reconciliation state:
     * - 'searching': Walking but not yet inside target subtree
     * - 'reconciling': Inside target subtree, applying changes
     * - 'done': Exited target subtree, no more changes
     */
    private reconcileState: 'searching' | 'reconciling' | 'done' = 'reconciling';

    /**
     * Reconciles whitespace from a formatted tree into the original tree.
     *
     * @param original The original tree (with types, markers, etc.)
     * @param formatted The formatted tree (with desired whitespace)
     * @param targetSubtree Optional subtree to limit reconciliation to. If provided,
     *                      only this subtree and its descendants will have whitespace applied.
     * @returns The original tree with whitespace from the formatted tree
     */
    async reconcile(original: J, formatted: J, targetSubtree?: J): Promise<J> {
        this.compatible = true;
        this.formattedCursor = undefined;
        this.cursor = new Cursor(undefined, undefined);
        this.targetSubtreeId = targetSubtree?.id;
        this.reconcileState = targetSubtree ? 'searching' : 'reconciling';

        const result = await this.visit(original, formatted);
        return result ?? original;
    }

    /**
     * Check if we should apply whitespace changes at the current position.
     */
    private shouldReconcile(): boolean {
        return this.reconcileState === 'reconciling';
    }

    /**
     * Marks structure as incompatible and returns the original unchanged.
     */
    protected structureMismatch<T>(t: T): T {
        this.compatible = false;
        return t;
    }

    /**
     * Copies a Space from formatted tree. This is where the actual whitespace
     * reconciliation happens.
     *
     * Preserves the original whitespace if the formatted version would remove newlines,
     * as this indicates a structural change (e.g., collapsing multi-line to single-line).
     */
    protected reconcileSpace(original: J.Space, formatted: J.Space): J.Space {
        // Only apply whitespace changes when inside the target subtree
        if (!this.shouldReconcile()) {
            return original;
        }

        // Preserve original if formatted would remove newlines (structural change)
        const originalHasNewline = original.whitespace.includes('\n');
        const formattedHasNewline = formatted.whitespace.includes('\n');
        if (originalHasNewline && !formattedHasNewline) {
            return original;
        }

        // Return the formatted space - this is the core of reconciliation
        return formatted;
    }

    /**
     * Visit a property value, handling all the different types appropriately.
     */
    protected async visitProperty(original: any, formatted: any): Promise<any> {
        // Handle null/undefined
        if (original == null || formatted == null) {
            if (original !== formatted) {
                return this.structureMismatch(original);
            }
            return original;
        }

        const kind = (original as any).kind;

        // Space - copy from formatted
        if (kind === J.Kind.Space) {
            return this.reconcileSpace(original, formatted);
        }

        // Type nodes - short-circuit, keep original (types don't have whitespace)
        if (typeof kind === 'string' && kind.startsWith('org.openrewrite.java.tree.JavaType$')) {
            return original;
        }

        // RightPadded wrapper
        if (kind === J.Kind.RightPadded) {
            return this.visitRightPaddedReconcile(original, formatted);
        }

        // LeftPadded wrapper
        if (kind === J.Kind.LeftPadded) {
            return this.visitLeftPaddedReconcile(original, formatted);
        }

        // Container wrapper
        if (kind === J.Kind.Container) {
            return this.visitContainerReconcile(original, formatted);
        }

        // Tree node (has a kind property)
        if (kind !== undefined && typeof kind === 'string') {
            return this.visit(original, formatted);
        }

        // Primitive values - return original unchanged
        return original;
    }

    /**
     * Visit all properties of an element and copy prefix from formatted.
     */
    protected async visitElement<T extends J>(original: T, formatted: T): Promise<T> {
        if (!this.compatible) return original;

        // Check if kinds match
        if (original.kind !== formatted.kind) {
            return this.structureMismatch(original);
        }

        // Start with original, will copy prefix
        let result = original;

        // Copy prefix from formatted only when reconciling, and preserve original if newlines would be removed
        if (this.shouldReconcile() && 'prefix' in original && 'prefix' in formatted) {
            const originalPrefix = (original as any).prefix as J.Space;
            const formattedPrefix = (formatted as any).prefix as J.Space;
            const originalHasNewline = originalPrefix.whitespace.includes('\n');
            const formattedHasNewline = formattedPrefix.whitespace.includes('\n');
            if (!originalHasNewline || formattedHasNewline) {
                result = produce(result, (draft: any) => {
                    draft.prefix = formattedPrefix;
                });
            }
        }

        // Visit all child properties
        for (const key of Object.keys(original)) {
            // Skip: kind, id, markers, prefix (already copied), type properties
            if (key === 'kind' || key === 'id' || key === 'markers' || key === 'prefix' ||
                key === 'type' || key === 'fieldType' || key === 'variableType' ||
                key === 'methodType' || key === 'constructorType') {
                continue;
            }

            const originalValue = (original as any)[key];
            const formattedValue = (formatted as any)[key];

            // Handle arrays
            if (Array.isArray(originalValue)) {
                if (!Array.isArray(formattedValue) || originalValue.length !== formattedValue.length) {
                    return this.structureMismatch(original);
                }

                const newArray: any[] = [];
                let changed = false;
                for (let i = 0; i < originalValue.length; i++) {
                    const visited = await this.visitProperty(originalValue[i], formattedValue[i]);
                    if (!this.compatible) return original;
                    newArray.push(visited);
                    if (visited !== originalValue[i]) {
                        changed = true;
                    }
                }

                if (changed) {
                    result = produce(result, (draft: any) => {
                        draft[key] = newArray;
                    });
                }
            } else {
                // Visit the property
                const visited = await this.visitProperty(originalValue, formattedValue);
                if (!this.compatible) return original;

                if (visited !== originalValue) {
                    result = produce(result, (draft: any) => {
                        draft[key] = visited;
                    });
                }
            }
        }

        return result;
    }

    /**
     * Override visit to use visitElement for all nodes.
     */
    override async visit<R extends J>(tree: Tree, p: J, parent?: Cursor): Promise<R | undefined> {
        if (!this.compatible) return tree as R;

        const original = tree as J;
        const formatted = p;

        // Check kind match
        if (original.kind !== formatted.kind) {
            return this.structureMismatch(original) as R;
        }

        // Track entering the target subtree
        const isTargetSubtree = this.targetSubtreeId !== undefined && original.id === this.targetSubtreeId;
        const previousState = this.reconcileState;
        if (isTargetSubtree && this.reconcileState === 'searching') {
            this.reconcileState = 'reconciling';
        }

        // Update formattedCursor
        const savedFormattedCursor = this.formattedCursor;
        this.formattedCursor = new Cursor(formatted, this.formattedCursor);

        try {
            const result = await this.visitElement(original, formatted);
            return result as R;
        } finally {
            this.formattedCursor = savedFormattedCursor;
            // Track exiting the target subtree
            if (isTargetSubtree && previousState === 'searching') {
                this.reconcileState = 'done';
            }
        }
    }

    /**
     * Reconcile RightPadded - copy 'after' whitespace.
     */
    protected async visitRightPaddedReconcile<T extends J | boolean>(
        original: J.RightPadded<T>,
        formatted: J.RightPadded<T>
    ): Promise<J.RightPadded<T>> {
        if (!this.compatible) return original;

        if ((formatted as any).kind !== J.Kind.RightPadded) {
            return this.structureMismatch(original);
        }

        // Visit the element
        const visitedElement = await this.visitProperty(original.element, formatted.element);
        if (!this.compatible) return original;

        // Copy 'after' whitespace only when reconciling, and preserve original if newlines would be removed
        return produce(original, draft => {
            (draft as any).element = visitedElement;
            if (this.shouldReconcile()) {
                const originalHasNewline = original.after.whitespace.includes('\n');
                const formattedHasNewline = formatted.after.whitespace.includes('\n');
                if (!originalHasNewline || formattedHasNewline) {
                    draft.after = formatted.after;
                }
            }
        });
    }

    /**
     * Reconcile LeftPadded - copy 'before' whitespace.
     */
    protected async visitLeftPaddedReconcile<T extends J | J.Space | number | string | boolean>(
        original: J.LeftPadded<T>,
        formatted: J.LeftPadded<T>
    ): Promise<J.LeftPadded<T>> {
        if (!this.compatible) return original;

        if ((formatted as any).kind !== J.Kind.LeftPadded) {
            return this.structureMismatch(original);
        }

        // Visit the element
        const visitedElement = await this.visitProperty(original.element, formatted.element);
        if (!this.compatible) return original;

        // Copy 'before' whitespace only when reconciling, and preserve original if newlines would be removed
        return produce(original, draft => {
            (draft as any).element = visitedElement;
            if (this.shouldReconcile()) {
                const originalHasNewline = original.before.whitespace.includes('\n');
                const formattedHasNewline = formatted.before.whitespace.includes('\n');
                if (!originalHasNewline || formattedHasNewline) {
                    draft.before = formatted.before;
                }
            }
        });
    }

    /**
     * Reconcile Container - copy 'before' whitespace and visit elements.
     */
    protected async visitContainerReconcile<T extends J>(
        original: J.Container<T>,
        formatted: J.Container<T>
    ): Promise<J.Container<T>> {
        if (!this.compatible) return original;

        if ((formatted as any).kind !== J.Kind.Container) {
            return this.structureMismatch(original);
        }

        // Check array length
        if (original.elements.length !== formatted.elements.length) {
            return this.structureMismatch(original);
        }

        // Visit each element
        const newElements: J.RightPadded<T>[] = [];
        for (let i = 0; i < original.elements.length; i++) {
            const visited = await this.visitRightPaddedReconcile(
                original.elements[i],
                formatted.elements[i]
            );
            if (!this.compatible) return original;
            newElements.push(visited);
        }

        // Copy 'before' whitespace only when reconciling, and preserve original if newlines would be removed
        return produce(original, draft => {
            if (this.shouldReconcile()) {
                const originalHasNewline = original.before.whitespace.includes('\n');
                const formattedHasNewline = formatted.before.whitespace.includes('\n');
                if (!originalHasNewline || formattedHasNewline) {
                    draft.before = formatted.before;
                }
            }
            (draft as any).elements = newElements;
        });
    }
}

File: rewrite-csharp/csharp/OpenRewrite/CSharp/Format/WhitespaceReconciler.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System.Collections;
using System.Collections.Concurrent;
using System.Reflection;
using OpenRewrite.Core;
using OpenRewrite.Java;

namespace OpenRewrite.CSharp.Format;

/// <summary>
/// Walks two structurally identical trees in parallel, copying Space and Markers
/// from the formatted tree to the original. Preserves the original tree's IDs,
/// types, and all non-whitespace state.
///
/// If the trees diverge structurally, returns the original unchanged and sets
/// <see cref="IsCompatible"/> to false.
/// </summary>
public class WhitespaceReconciler
{
    private static readonly ConcurrentDictionary<Type, PropertyInfo[]> PropertyCache = new();
    private static readonly ConcurrentDictionary<(Type, string), MethodInfo?> WithMethodCache = new();

    private static readonly HashSet<string> SkipProperties =
        ["Id", "SourcePath"];

    public static bool ThrowOnMismatchDefault { get; set; }

    private const int MaxMismatches = 5;
    private readonly bool _throwOnMismatch = ThrowOnMismatchDefault;
    private List<MismatchEntry>? _mismatches;
    private readonly Stack<string> _path = new();

    private bool _compatible = true;
    private J? _targetSubtree;
    private J? _stopAfter;
    private HashSet<Guid>? _targetIds;
    private int _multiTargetDepth;
    private ReconcileState _state;

    private enum ReconcileState
    {
        Searching,
        Reconciling,
        Done
    }

    public bool IsCompatible => _compatible;

    public J Reconcile(J original, J formatted, J? targetSubtree = null, J? stopAfter = null)
    {
        _compatible = true;
        _targetSubtree = targetSubtree;
        _stopAfter = stopAfter;
        _targetIds = null;
        _state = targetSubtree != null ? ReconcileState.Searching : ReconcileState.Reconciling;

        var result = VisitTree(original, formatted);
        return result as J ?? original;
    }

    /// <summary>
    /// Reconcile whitespace only within subtrees whose root IDs are in <paramref name="targetIds"/>.
    /// Nodes outside these subtrees keep their original whitespace.
    /// </summary>
    public J Reconcile(J original, J formatted, HashSet<Guid> targetIds)
    {
        _compatible = true;
        _targetSubtree = null;
        _stopAfter = null;
        _targetIds = targetIds;
        _multiTargetDepth = 0;
        _state = ReconcileState.Searching;

        var result = VisitTree(original, formatted);
        return result as J ?? original;
    }

    private bool ShouldReconcile() => _state == ReconcileState.Reconciling;

    private void PushPath(string segment) { if (_throwOnMismatch) _path.Push(segment); }
    private void PopPath() { if (_throwOnMismatch) _path.Pop(); }

    private string CurrentPath()
    {
        return string.Join(".", _path.Reverse());
    }

    private object? StructureMismatch(object? original, object? formatted = null)
    {
        _compatible = false;
        if (_throwOnMismatch)
        {
            _mismatches ??= new List<MismatchEntry>(MaxMismatches);
            if (_mismatches.Count < MaxMismatches)
            {
                _mismatches.Add(new MismatchEntry(
                    CurrentPath(),
                    original?.GetType().Name ?? "null",
                    formatted?.GetType().Name ?? "null"));
            }
        }
        return original;
    }

    private object? VisitProperty(object? original, object? formatted)
    {
        if (!_compatible && !_throwOnMismatch) return original;

        // Handle null: if one is null and the other isn't, check whether it's a
        // structural type (J, padded wrapper, list) where a mismatch is fatal.
        // For primitive/string properties (e.g., Modifier.Keyword), null vs non-null
        // is not a structural divergence — just keep the original value.
        if (original == null || formatted == null)
        {
            if (ReferenceEquals(original, formatted))
                return original;

            // A null mismatch on a structural type is a real incompatibility
            var nonNull = original ?? formatted;
            if (nonNull is J || nonNull is Space || nonNull is Markers ||
                nonNull is IList ||
                IsGenericOf(nonNull, typeof(JRightPadded<>)) ||
                IsGenericOf(nonNull, typeof(JLeftPadded<>)) ||
                IsGenericOf(nonNull, typeof(JContainer<>)))
            {
                return StructureMismatch(original, formatted);
            }

            // For primitive types (string, enum, etc.), keep the original
            return original;
        }

        // Space — copy from formatted when reconciling
        if (original is Space)
        {
            if (!ShouldReconcile()) return original;
            return formatted is Space ? formatted : StructureMismatch(original, formatted);
        }

        // Markers — copy from formatted when reconciling
        if (original is Markers)
        {
            if (!ShouldReconcile()) return original;
            return formatted is Markers ? formatted : StructureMismatch(original, formatted);
        }

        // J nodes — recurse
        if (original is J origJ)
        {
            if (formatted is not J fmtJ)
                return StructureMismatch(original, formatted);
            return VisitTree(origJ, fmtJ);
        }

        // Padded wrappers — recurse into their structure
        if (IsGenericOf(original, typeof(JRightPadded<>)))
            return VisitRightPadded(original, formatted);

        if (IsGenericOf(original, typeof(JLeftPadded<>)))
            return VisitLeftPadded(original, formatted);

        if (IsGenericOf(original, typeof(JContainer<>)))
            return VisitContainer(original, formatted);

        // Lists (of padded elements, child nodes, etc.)
        if (original is IList origList)
        {
            if (formatted is not IList fmtList)
                return StructureMismatch(original, formatted);
            return VisitList(origList, fmtList);
        }

        // Primitive values, enums, etc. — keep original
        return original;
    }

    private object? VisitTree(J original, J formatted)
    {
        if (!_compatible && !_throwOnMismatch) return original;

        // Check structural type compatibility
        if (original.GetType() != formatted.GetType())
            return StructureMismatch(original, formatted);

        // Track target subtree (single-target mode)
        var isTarget = _targetSubtree != null && ReferenceEquals(original, _targetSubtree);
        var isStopAfter = _stopAfter != null && ReferenceEquals(original, _stopAfter);

        // Track target IDs (multi-target mode): enter Reconciling when we hit a target ID
        var isMultiTarget = _targetIds != null && _state == ReconcileState.Searching &&
                            _targetIds.Contains(original.Id);

        var previousState = _state;

        if (isTarget && _state == ReconcileState.Searching)
            _state = ReconcileState.Reconciling;
        if (isMultiTarget)
        {
            _state = ReconcileState.Reconciling;
            _multiTargetDepth = 0;
        }
        if (_targetIds != null && _state == ReconcileState.Reconciling)
            _multiTargetDepth++;

        PushPath(original.GetType().Name);
        try
        {
            var type = original.GetType();
            var properties = GetReconcilableProperties(type);
            var result = original;

            foreach (var prop in properties)
            {
                PushPath(prop.Name);
                var origVal = prop.GetValue(original);
                var fmtVal = prop.GetValue(formatted);

                var visited = VisitProperty(origVal, fmtVal);
                PopPath();
                if (!_compatible && !_throwOnMismatch) return original;

                if (!ReferenceEquals(visited, origVal))
                {
                    result = SetProperty(result, type, prop, visited);
                    if (result == null) return original;
                }
            }

            return result;
        }
        finally
        {
            PopPath();
            if (isTarget && previousState == ReconcileState.Searching)
                _state = ReconcileState.Done;
            if (isStopAfter && previousState == ReconcileState.Reconciling)
                _state = ReconcileState.Done;

            // Multi-target: return to Searching when leaving a target subtree
            if (_targetIds != null && _state == ReconcileState.Reconciling)
            {
                _multiTargetDepth--;
                if (_multiTargetDepth <= 0)
                    _state = ReconcileState.Searching;
            }
        }
    }

    private object? VisitRightPadded(object original, object formatted)
    {
        if (!_compatible && !_throwOnMismatch) return original;

        var origType = original.GetType();
        var fmtType = formatted.GetType();
        if (origType != fmtType) return StructureMismatch(original, formatted);

        var elementProp = origType.GetProperty("Element")!;
        var afterProp = origType.GetProperty("After")!;
        var markersProp = origType.GetProperty("Markers")!;

        var origElement = elementProp.GetValue(original);
        var fmtElement = elementProp.GetValue(formatted);
        PushPath("Element");
        var visitedElement = VisitProperty(origElement, fmtElement);
        PopPath();
        if (!_compatible && !_throwOnMismatch) return original;

        var origAfter = afterProp.GetValue(original) as Space;
        var fmtAfter = afterProp.GetValue(formatted) as Space;
        PushPath("After");
        var visitedAfter = ShouldReconcile() ? fmtAfter : origAfter;
        PopPath();

        var origMarkers = markersProp.GetValue(original) as Markers;
        var fmtMarkers = markersProp.GetValue(formatted) as Markers;
        PushPath("Markers");
        var visitedMarkers = ShouldReconcile() ? fmtMarkers : origMarkers;
        PopPath();

        if (ReferenceEquals(visitedElement, origElement) &&
            ReferenceEquals(visitedAfter, origAfter) &&
            ReferenceEquals(visitedMarkers, origMarkers))
            return original;

        var withElement = origType.GetMethod("WithElement")!;
        var withAfter = origType.GetMethod("WithAfter")!;
        var withMarkers = origType.GetMethod("WithMarkers")!;

        var result = original;
        if (!ReferenceEquals(visitedElement, origElement))
            result = withElement.Invoke(result, [visitedElement])!;
        if (!ReferenceEquals(visitedAfter, origAfter))
            result = withAfter.Invoke(result, [visitedAfter])!;
        if (!ReferenceEquals(visitedMarkers, origMarkers))
            result = withMarkers.Invoke(result, [visitedMarkers])!;

        return result;
    }

    private object? VisitLeftPadded(object original, object formatted)
    {
        if (!_compatible && !_throwOnMismatch) return original;

        var origType = original.GetType();
        var fmtType = formatted.GetType();
        if (origType != fmtType) return StructureMismatch(original, formatted);

        var beforeProp = origType.GetProperty("Before")!;
        var elementProp = origType.GetProperty("Element")!;

        var origBefore = beforeProp.GetValue(original) as Space;
        var fmtBefore = beforeProp.GetValue(formatted) as Space;
        PushPath("Before");
        var visitedBefore = ShouldReconcile() ? fmtBefore : origBefore;
        PopPath();

        var origElement = elementProp.GetValue(original);
        var fmtElement = elementProp.GetValue(formatted);
        PushPath("Element");
        var visitedElement = VisitProperty(origElement, fmtElement);
        PopPath();
        if (!_compatible && !_throwOnMismatch) return original;

        if (ReferenceEquals(visitedBefore, origBefore) &&
            ReferenceEquals(visitedElement, origElement))
            return original;

        var withBefore = origType.GetMethod("WithBefore")!;
        var withElement = origType.GetMethod("WithElement")!;

        var result = original;
        if (!ReferenceEquals(visitedBefore, origBefore))
            result = withBefore.Invoke(result, [visitedBefore])!;
        if (!ReferenceEquals(visitedElement, origElement))
            result = withElement.Invoke(result, [visitedElement])!;

        return result;
    }

    private object? VisitContainer(object original, object formatted)
    {
        if (!_compatible && !_throwOnMismatch) return original;

        var origType = original.GetType();
        var fmtType = formatted.GetType();
        if (origType != fmtType) return StructureMismatch(original, formatted);

        var beforeProp = origType.GetProperty("Before")!;
        var elementsProp = origType.GetProperty("Elements")!;
        var markersProp = origType.GetProperty("Markers")!;

        var origBefore = beforeProp.GetValue(original) as Space;
        var fmtBefore = beforeProp.GetValue(formatted) as Space;
        PushPath("Before");
        var visitedBefore = ShouldReconcile() ? fmtBefore : origBefore;
        PopPath();

        var origElements = elementsProp.GetValue(original) as IList;
        var fmtElements = elementsProp.GetValue(formatted) as IList;
        PushPath("Elements");
        var visitedElements = VisitList(origElements!, fmtElements!);
        PopPath();
        if (!_compatible && !_throwOnMismatch) return original;

        var origMarkers = markersProp.GetValue(original) as Markers;
        var fmtMarkers = markersProp.GetValue(formatted) as Markers;
        PushPath("Markers");
        var visitedMarkers = ShouldReconcile() ? fmtMarkers : origMarkers;
        PopPath();

        if (ReferenceEquals(visitedBefore, origBefore) &&
            ReferenceEquals(visitedElements, origElements) &&
            ReferenceEquals(visitedMarkers, origMarkers))
            return original;

        var withBefore = origType.GetMethod("WithBefore")!;
        var withElements = origType.GetMethod("WithElements")!;
        var withMarkers = origType.GetMethod("WithMarkers")!;

        var result = original;
        if (!ReferenceEquals(visitedBefore, origBefore))
            result = withBefore.Invoke(result, [visitedBefore])!;
        if (!ReferenceEquals(visitedElements, origElements))
            result = withElements.Invoke(result, [visitedElements])!;
        if (!ReferenceEquals(visitedMarkers, origMarkers))
            result = withMarkers.Invoke(result, [visitedMarkers])!;

        return result;
    }

    private object? VisitList(IList original, IList formatted)
    {
        if (original.Count != formatted.Count)
            return StructureMismatch(original, formatted);

        var changed = false;
        var newList = new List<object?>(original.Count);

        for (var i = 0; i < original.Count; i++)
        {
            PushPath($"[{i}]");
            var visited = VisitProperty(original[i], formatted[i]);
            PopPath();
            if (!_compatible && !_throwOnMismatch) return original;
            newList.Add(visited);
            if (!ReferenceEquals(visited, original[i]))
                changed = true;
        }

        if (!changed) return original;

        // Create a new typed list matching the original's element type
        var listType = original.GetType();
        if (listType.IsGenericType)
        {
            var elementType = listType.GetGenericArguments()[0];
            var typedList = (IList)Activator.CreateInstance(typeof(List<>).MakeGenericType(elementType))!;
            foreach (var item in newList)
                typedList.Add(item);
            return typedList;
        }

        return newList;
    }

    private static bool IsJavaType(Type type)
    {
        var t = Nullable.GetUnderlyingType(type) ?? type;
        return typeof(JavaType).IsAssignableFrom(t);
    }

    private static PropertyInfo[] GetReconcilableProperties(Type type)
    {
        return PropertyCache.GetOrAdd(type, t =>
            t.GetProperties(BindingFlags.Public | BindingFlags.Instance)
                .Where(p => p.CanRead && !SkipProperties.Contains(p.Name) && !IsJavaType(p.PropertyType))
                .ToArray());
    }

    private static J? SetProperty(J node, Type type, PropertyInfo prop, object? value)
    {
        var key = (type, prop.Name);
        var withMethod = WithMethodCache.GetOrAdd(key, k =>
        {
            // Look for With{PropertyName} method
            var methodName = $"With{k.Item2}";
            return k.Item1.GetMethod(methodName, [prop.PropertyType]);
        });

        if (withMethod == null)
        {
            // No With* method — can't set this property, skip
            return node;
        }

        return withMethod.Invoke(node, [value]) as J;
    }

    private static bool IsGenericOf(object obj, Type genericTypeDef)
    {
        var type = obj.GetType();
        return type.IsGenericType && type.GetGenericTypeDefinition() == genericTypeDef;
    }

    public record MismatchEntry(string Path, string OriginalType, string FormattedType)
    {
        public override string ToString() => $"{Path}: {OriginalType} vs {FormattedType}";
    }

    public class WhitespaceReconcileMismatchException(IReadOnlyList<MismatchEntry> mismatches)
        : Exception(FormatMessage(mismatches))
    {
        public IReadOnlyList<MismatchEntry> Mismatches { get; } = mismatches;

        private static string FormatMessage(IReadOnlyList<MismatchEntry> mismatches)
        {
            var sb = new System.Text.StringBuilder();
            sb.AppendLine($"{mismatches.Count} structural mismatch(es) between original and formatted trees:");
            for (var i = 0; i < mismatches.Count; i++)
                sb.AppendLine($"  {i + 1}. {mismatches[i]}");
            return sb.ToString();
        }
    }
}
